README.md
# multisax
A module which allows multiple SAX library transparently.

[![Build Status](https://travis-ci.org/cielavenir/multisax.png)](https://travis-ci.org/cielavenir/multisax) [![Code Climate](https://codeclimate.com/github/cielavenir/multisax.png)](https://codeclimate.com/github/cielavenir/multisax) [![Coverage Status](https://coveralls.io/repos/cielavenir/multisax/badge.png)](https://coveralls.io/r/cielavenir/multisax)

[Why the coverage is so low?](Xerces.md)

## Supported Ruby versions
* Ruby 1.8.7 or later
* jruby
  * libxml/ox/xmlparser are not available (they use C ext).
* rubinius 2.1.1 or 2.2.3 (or later, maybe)
  * 2.2.1 and 2.2.2 are not working.
  * xmlparser is not available.
* (Possibly) ironruby / macruby / topaz etc

## Binary distribution
* https://rubygems.org/gems/multisax

## (Embeddable) Minimalistic Edition
* https://gist.github.com/cielavenir/7691221
  * multisax_mini.rb:  :libxml/:rexmlstream/:rexmlsax2
  * multisax_mini2.rb: :libxml/:rexmlstream

## Benchmark
* https://gist.github.com/cielavenir/8398647

## Install
* gem install multisax
* Optional XML libraries:
* gem install ox
* gem install libxml-ruby
* gem install nokogiri
* gem install xmlparser
  * to treat cdata, I recommend expat 2.x.
  * For example, OSX MacPorts users should:
  * gem install xmlparser -- --with-expat-include=/opt/local/include --with-expat-lib=/opt/local/lib
* gem install oga

## Usage
* Please check spec/multisax.spec as an example.
* Complex usage:

```rb
require 'multisax'
listener=MultiSAX::Sax.parse(xml,Class.new{
	include MultiSAX::Callbacks
	def initialize
		@content=Hash.new{|h,k|h[k]=[]}
		@current_tag=[]
	end
	attr_reader :content

	def sax_tag_start(tag,attrs)
		@current_tag.push(tag)
	end
	def sax_tag_end(tag)
		if (t=@current_tag.pop)!=tag then raise "xml is malformed /#{t}" end
	end
	def sax_cdata(text)
		@content[@current_tag.last] << text
	end
	def sax_text(text)
		text.strip!
		@content[@current_tag.last] << text if text.size>0
	end
	def sax_comment(text)
	end
}.new)
listener.content.each{...}
```

## Contributing to multisax
* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

## Copyright
Copyright (c) 2014 T. Yamada under Ruby License (2-clause BSDL or Artistic).
See LICENSE.txt for further details.


CHANGELOG.md
# ChangeLog

## 0.0.7.1 (2015 Nov 24)
- Forgot to add date to changelog

## 0.0.7 (2015 Nov 24)
- Update oga wrapper.

## 0.0.6 (2014 Nov 19)
- Now supports oga >= 0.2.0.

## 0.0.5 (2014 Oct 7)
- Added oga gem support (Ruby 1.9.x or later required).
- Integrated Coverall.

## 0.0.4.2 (2014 Jan 14)
- Fixed regression in parsefile().

## 0.0.4.1 (2014 Jan 14)
- Fixed small stuff.

## 0.0.4 (2014 Jan 13)
- Integrated travis. Now Ruby/Rubinius/jruby supports are assured.
- Added xmlparser (expat bindings) gem support.

## 0.0.3 (2013 Nov 14)
- Fixed namespace handling.
- Now you can also select :oxhtml to parse HTML.
- Added shortcut :XML and :HTML.
- sax_tag_start()'s attrs is assured to be a Hash.
- Refined spec.

## 0.0.2 (2013 Nov 13)
- Now you can create an instance of MultiSAX::SAX.
  - Please note that passed class to MultiSAX is still modified directly.
  - So only MultiSAX::SAX instances are thread-safe.
  - MultiSAX::Sax is now an instance of MultiSAX::SAX (this is backward-compatible).
- You can specify :nokogirihtml explicitly to parse HTML.
- Fixed attrs with Ox (now String is passed, not Symbol)
- Moved to Bundler rather than Jeweler.

## 0.0.1 (2013 Jul 8)
- Added ChangeLog.
- Added Ruby 1.8.7 support.
  - Might work on lower version, but not guaranteed.
  - JRuby etc's supports depend on "require successfulness".
  - Nokogiri is supported up to 1.5.x on Ruby 1.8.x.
- MultiSAX::Sax.parse supports IO.
  - rexmlsax2's stringio support is limited on Ruby 1.8.x. Be careful.
  - Unless you directly specifies it, usually rexmlstream is selected.
- Added MultiSAX::Sax.parsefile().

## 0.0.0.5 (2013 Jun 21)
- Added a test case.

## 0.0.0.4 (2013 Jun 15)
- Provided rdoc.

## 0.0.0.3 (2013 Jun 14)
- Fixed @@parser getter.

## 0.0.0.2 (2013 Jun 14)
- Fixed Rakefile.

## 0.0.0.1 (2013 Jun 14)
- First release.


spec/spec_helper.rb
require 'rspec'
RSpec.configure{|config|
	config.color=true
}

require 'stringio'
if RUBY_VERSION<'1.9' #gee, :rexmlsax2 on Ruby 1.8 dislikes StringIO.
	class StringIO
		def stat
			Class.new{
				def pipe?() return false end
			}.new
		end
	end
end

if !defined?(RUBY_ENGINE)||RUBY_ENGINE=='ruby'
	require 'simplecov'
	require 'coveralls'
	SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
		SimpleCov::Formatter::HTMLFormatter,
		Coveralls::SimpleCov::Formatter
	]
	SimpleCov.start do
		add_filter 'spec'
	end
end

require File.expand_path(File.dirname(__FILE__)+'/../lib/multisax')

spec/multisax_spec.rb
require File.expand_path(File.dirname(__FILE__)+'/spec_helper')

class MultiSAXTester
	include MultiSAX::Callbacks
	def initialize
		@result=[]
		@attrib=nil
	end
	def sax_tag_start(tag,attrs)
		@result<<tag
		@attrib=attrs['class'] if tag=='span'
	end
	def sax_tag_end(tag)
		@result<<tag
	end
	def sax_text(text)
		text.strip!
		@result<<text if text.size>0
	end
	def sax_cdata(text)
		text.strip!
		@result<<text if text.size>0
	end
	def sax_xmldecl(version,encoding,standalone)
		@xmlencoding=encoding
	end
	attr_reader :result,:attrib,:xmlencoding
end

input_xml=<<"EOM"
<?xml version="1.0" encoding="UTF-8"?>
<ns xmlns:zzz="http://example.com/">
<zzz:hello><![CDATA[sax$]]><span class="foo">world</span></zzz:hello>
</ns>
EOM
xml_answer=['ns','zzz:hello','sax$','span','world','span','zzz:hello','ns']

describe "[XML] MultiSAX::Sax.parse(String)" do
	it "fails on :unknown" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:unknown).should be_falsey
	end
	it "uses :rexmlstream" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:rexmlstream)
		MultiSAX::Sax.parser.should eq :rexmlstream
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :rexmlsax2" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:rexmlsax2)
		MultiSAX::Sax.parser.should eq :rexmlsax2
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :ox" do
		pending 'ox is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'ox looks broken on Ruby <1.9' if RUBY_VERSION<'1.9' ###
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:ox)
		MultiSAX::Sax.parser.should eq :ox
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :libxml" do
		pending 'libxml is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:libxml)
		MultiSAX::Sax.parser.should eq :libxml
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		#listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :nokogiri" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:nokogiri)
		MultiSAX::Sax.parser.should eq :nokogiri
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :xmlparser" do
		pending 'xmlparser is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'xmlparser is not supported by rubinius' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='rbx'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:xmlparser)
		MultiSAX::Sax.parser.should eq :xmlparser
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :oga" do
		pending 'oga requires Ruby >=1.9.3' if RUBY_VERSION<'1.9.3'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:oga)
		MultiSAX::Sax.parser.should eq :oga
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :xerces" do
		pending 'xerces is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'xerces will not be supported officially'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:xerces)
		MultiSAX::Sax.parser.should eq :xerces
		listener=MultiSAX::Sax.parse(input_xml,MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		#listener.xmlencoding.should eq 'UTF-8'
	end
end

describe "[XML] MultiSAX::Sax.parse(IO)" do
	it "uses :rexmlstream" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:rexmlstream)
		MultiSAX::Sax.parser.should eq :rexmlstream
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :rexmlsax2" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:rexmlsax2)
		MultiSAX::Sax.parser.should eq :rexmlsax2
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :ox" do
		pending 'ox is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'ox looks broken on Ruby <1.9' if RUBY_VERSION<'1.9' ###
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:ox)
		MultiSAX::Sax.parser.should eq :ox
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :libxml" do
		pending 'libxml is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:libxml)
		MultiSAX::Sax.parser.should eq :libxml
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		#listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :nokogiri" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:nokogiri)
		MultiSAX::Sax.parser.should eq :nokogiri
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :xmlparser" do
		pending 'xmlparser is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'xmlparser is not supported by rubinius' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='rbx'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:xmlparser)
		MultiSAX::Sax.parser.should eq :xmlparser
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :oga" do
		pending 'oga requires Ruby >=1.9.3' if RUBY_VERSION<'1.9.3'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:oga)
		MultiSAX::Sax.parser.should eq :oga
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		listener.xmlencoding.should eq 'UTF-8'
	end
	it "uses :xerces" do
		pending 'xerces is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'xerces will not be supported officially'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:xerces)
		MultiSAX::Sax.parser.should eq :xerces
		listener=MultiSAX::Sax.parse(StringIO.new(input_xml),MultiSAXTester.new)
		listener.result.should eq xml_answer
		listener.attrib.should eq 'foo'
		#listener.xmlencoding.should eq 'UTF-8'
	end
end

# broken intentionally
# nokogiri-java wants head tag...
input_html=<<"EOM"
<html>
<head></head>
<body>
<span class="foo">hello
</body>
</html>
EOM
html_answer=['html','head','head','body','span','hello','span','body','html']

describe "[HTML] MultiSAX::Sax.parse(String)" do
	it "uses :oxhtml" do
		pending 'ox is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'ox looks broken on Ruby <1.9' if RUBY_VERSION<'1.9' ###
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:oxhtml)
		MultiSAX::Sax.parser.should eq :oxhtml
		listener=MultiSAX::Sax.parse(input_html,MultiSAXTester.new)
		listener.result.should eq html_answer
		listener.attrib.should eq 'foo'
	end
	it "uses :nokogirihtml" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:nokogirihtml)
		MultiSAX::Sax.parser.should eq :nokogirihtml
		listener=MultiSAX::Sax.parse(input_html,MultiSAXTester.new)
		listener.result.should eq html_answer
		listener.attrib.should eq 'foo'
	end
	it "uses :ogahtml" do
		pending 'oga requires Ruby >=1.9.3' if RUBY_VERSION<'1.9.3'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:ogahtml)
		MultiSAX::Sax.parser.should eq :ogahtml
		listener=MultiSAX::Sax.parse(input_html,MultiSAXTester.new)
		listener.result.should eq html_answer
		listener.attrib.should eq 'foo'
	end
end

describe "[HTML] MultiSAX::Sax.parse(IO)" do
	it "uses :oxhtml" do
		pending 'ox is not supported by jruby' if defined?(RUBY_ENGINE)&&RUBY_ENGINE=='jruby'
		pending 'ox looks broken on Ruby <1.9' if RUBY_VERSION<'1.9' ###
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:oxhtml)
		MultiSAX::Sax.parser.should eq :oxhtml
		listener=MultiSAX::Sax.parse(StringIO.new(input_html),MultiSAXTester.new)
		listener.result.should eq html_answer
		listener.attrib.should eq 'foo'
	end
	it "uses :nokogirihtml" do
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:nokogirihtml)
		MultiSAX::Sax.parser.should eq :nokogirihtml
		listener=MultiSAX::Sax.parse(StringIO.new(input_html),MultiSAXTester.new)
		listener.result.should eq html_answer
		listener.attrib.should eq 'foo'
	end
	it "uses :ogahtml" do
		pending 'oga requires Ruby >=1.9.3' if RUBY_VERSION<'1.9.3'
		MultiSAX::Sax.reset
		MultiSAX::Sax.open(:ogahtml)
		MultiSAX::Sax.parser.should eq :ogahtml
		listener=MultiSAX::Sax.parse(StringIO.new(input_html),MultiSAXTester.new)
		listener.result.should eq html_answer
		listener.attrib.should eq 'foo'
	end
end


lib/multisax.rb
# MultiSAX: Ruby Gem to handle multiple SAX libraries
#
# Copyright (c) 2014, T. Yamada under Ruby License (2-clause BSDL or Artistic).
#
# Check LICENSE terms.
#
# Note: MIT License is also applicable if that compresses LICENSE file.

module MultiSAX
	# VERSION string
	VERSION='0.0.7.1'

	# Default XML parsers
	XML_PARSERS_DEFAULT=[:rexmlstream,:rexmlsax2]

	# Installable XML parsers
	XML_PARSERS_INSTALLABLE=[:ox,:libxml,:xmlparser,:nokogiri,:oga]

	# Installable HTML parsers
	HTML_PARSERS_INSTALLABLE=[:oxhtml,:nokogirihtml,:ogahtml]

	# The class to handle XML libraries.
	class SAX
		# constructor.
		# list cannot be passed directly, since you should check the retval of open().
		def initialize
			@parser=nil
		end
		# Library loader.
		# Arguments are list (or Array) of libraries.
		#  if list is empty or :XML, the following are searched (order by speed):
		#  :ox, :libxml, :xmlparser, :nokogiri, :oga, :rexmlstream, :rexmlsax2
		#  if list is :HTML, the following are searched (order by speed):
		#  :oxhtml, :nokogirihtml, :ogahtml
		#  You can also specify libraries individually.
		#  If multiple selected, MultiSAX will try the libraries one by one and use the first usable one.
		def open(*list)
			return @parser if @parser
			list=XML_PARSERS_INSTALLABLE+XML_PARSERS_DEFAULT if list.empty?||list==[:XML]
			list=HTML_PARSERS_INSTALLABLE if list==[:HTML]
			list.each{|e_module|
				case e_module
					when :ox,:oxhtml
						begin
							require 'ox'
							require 'stringio' #this should be standard module.
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new(::Ox::Sax){
							def __init__(obj)
								@obj=obj
								@saxwrapper_tag=nil
								@saxwrapper_attr={}
								self
							end
							def end_element(tag) @obj.sax_tag_end(tag.to_s) end
							def cdata(txt) @obj.sax_cdata(txt) end
							def text(txt) @obj.sax_text(txt) end
							def comment(txt) @obj.sax_comment(txt) end

							def start_element(tag)
								if @after_error
									@obj.sax_tag_start(tag.to_s,{})
									@after_error=false
								else
									# I hope provided Listener's sax_tag_start will NOT be used elsewhere.
									#alias :attrs_done :attrs_done_normal
									@saxwrapper_tag=tag
									@saxwrapper_attr={}
								end
							end
							def attr(name,str)
								@saxwrapper_attr[name.to_s]=str
							end
							def attrs_done_xmldecl
								@obj.sax_xmldecl(@saxwrapper_attr['version'],@saxwrapper_attr['encoding'],@saxwrapper_attr['standalone'])
							end
							def attrs_done_normal
								@obj.sax_tag_start(@saxwrapper_tag.to_s,@saxwrapper_attr)
							end
							def attrs_done
								@saxwrapper_tag ? attrs_done_normal : attrs_done_xmldecl
							end
							def error(s,i,j) @after_error=true if s.end_with?('closed but not opened') end
						}
						break
					when :libxml
						begin
							require 'libxml'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new{
							include ::LibXML::XML::SaxParser::Callbacks
							def __init__(obj)
								@obj=obj
								self
							end
							def on_start_element(tag,attrs) @obj.sax_tag_start(tag,attrs) end
							def on_end_element(tag) @obj.sax_tag_end(tag) end
							def on_characters(txt) @obj.sax_text(txt) end
							def on_cdata_block(txt) @obj.sax_cdata(txt) end
							def on_comment(txt) @obj.sax_comment(txt) end
							#actually unused
							def xmldecl(version,encoding,standalone) @obj.sax_xmldecl(version,encoding,standalone) end
						}
						break
					when :nokogiri,:nokogirihtml
						#nokogiri 1.5.x are supported on Ruby 1.8.7.
						#next if RUBY_VERSION<'1.9'
						begin
							require 'nokogiri'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new(::Nokogiri::XML::SAX::Document){
							def __init__(obj)
								@obj=obj
								self
							end
							def start_element(tag,attrs) @obj.sax_tag_start(tag,attrs.is_a?(Array) ? Hash[*attrs.flatten(1)] : attrs) end
							def end_element(tag) @obj.sax_tag_end(tag) end
							def characters(txt) @obj.sax_text(txt) end
							def cdata_block(txt) @obj.sax_cdata(txt) end
							def comment(txt) @obj.sax_comment(txt) end
							def xmldecl(version,encoding,standalone) @obj.sax_xmldecl(version,encoding,standalone) end
						}
						break
					when :xmlparser
						begin
							require 'xml/saxdriver'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new(::XML::Parser){
							def __init__(obj)
								@obj=obj
								@cdata=false
								self
							end
							def startElement(tag,attrs) @obj.sax_tag_start(tag,attrs) end
							def endElement(tag) @obj.sax_tag_end(tag) end
							def comment(txt) @obj.sax_comment(txt) end
							def xmlDecl(version,encoding,standalone) @obj.sax_xmldecl(version,encoding,standalone) end
							def character(txt)
								if @cdata
									@obj.sax_cdata(txt)
								else
									@obj.sax_text(txt)
								end
							end
							def startCdata
								@cdata=true
							end
							def endCdata
								@cdata=false
							end
						}
						break
					when :oga,:ogahtml
						next if RUBY_VERSION<'1.9'
						begin
							require 'oga'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new{
							def __init__(obj)
								@obj=obj
								self
							end
							def on_element(ns,tag,attrs)
								tag_name=(ns ? (ns+':') : '')+tag
								@obj.sax_tag_start(tag_name,Hash[*attrs.flatten(1)])
								return tag_name
							end
							def after_element(ns,tag)
								tag_name=(ns ? (ns+':') : '')+tag
								@obj.sax_tag_end(tag_name)
							end
							def on_text(txt) @obj.sax_text(txt) end
							def on_cdata(txt) @obj.sax_cdata(txt) end
							def on_comment(txt) @obj.sax_comment(txt) end
							def on_xml_decl(args)
								attrs=Hash[*args.flatten(1)]
								@obj.sax_xmldecl(attrs['version'],attrs['encoding'],attrs['standalone'])
							end
						}
						break
					# :nocov:
					when :xerces
						begin
							require 'Xerces'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new(::XercesR::DocumentHandler){
							def __init__(obj)
								@obj=obj
								#@cdata=false
								self
							end
							def startElement(tag,attrs) @obj.sax_tag_start(tag,Hash[*attrs.getLength.times.map{|i|[attrs.getName(i),attrs.getValue(i)]}.flatten(1)]) end
							def endElement(tag) @obj.sax_tag_end(tag) end
							def comment(txt) @obj.sax_comment(txt) end
							#def xmlDecl(version,encoding,standalone) @obj.sax_xmldecl(version,encoding,standalone) end
							#def notationDecl(name, publicId, systemId) end
							def characters(txt,len)
								#if @cdata
									@obj.sax_cdata(txt)
								#else
									@obj.sax_text(txt)
								#end
							end
							#def startCdata
							#	@cdata=true
							#end
							#def endCdata
							#	@cdata=false
							#end
						}
						break
					# :nocov:
					when :rexmlstream
						begin
							require 'rexml/parsers/baseparser'
							require 'rexml/parsers/streamparser'
							require 'rexml/streamlistener'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new{
							include ::REXML::StreamListener
							def __init__(obj)
								@obj=obj
								self
							end
							def tag_start(tag,attrs) @obj.sax_tag_start(tag,attrs) end
							def tag_end(tag) @obj.sax_tag_end(tag) end
							def text(txt) @obj.sax_text(txt) end
							def cdata(txt) @obj.sax_cdata(txt) end
							def comment(txt) @obj.sax_comment(txt) end
							def xmldecl(version,encoding,standalone) @obj.sax_xmldecl(version,encoding,standalone) end
						}
						break
					when :rexmlsax2
						begin
							require 'rexml/parsers/sax2parser'
							require 'rexml/sax2listener'
						rescue LoadError;next end
						@parser=e_module
						@saxhelper=Class.new{
							include ::REXML::SAX2Listener
							def __init__(obj)
								@obj=obj
								self
							end
							def start_element(uri,tag,qname,attrs) @obj.sax_tag_start(qname,attrs) end
							def end_element(uri,tag,qname) @obj.sax_tag_end(qname) end
							def characters(txt) @obj.sax_text(txt) end
							def cdata(txt) @obj.sax_cdata(txt) end
							def comment(txt) @obj.sax_comment(txt) end
							def xmldecl(version,encoding,standalone) @obj.sax_xmldecl(version,encoding,standalone) end
						}
						break
				end
			}
			return @parser
		end
		# Reset MultiSAX state so that you can re-open() another library.
		def reset() @parser=nil end
		# Returns which module is actually chosen.
		def parser() @parser end

		# The main parsing method.
		# Listener can be Class.new{include MultiSAX::Callbacks}.new. Returns the listener after SAX is applied.
		# If you have not called open(), this will call it using default value (all libraries).
		#  From 0.0.1, source can be IO as well as String.
		#  SAX's listeners are usually modified destructively.
		#  So instances shouldn't be provided.
		def parse(source,listener)
			if !@parser && !open
				raise "Failed to open SAX library. REXML, which is a standard Ruby module, might be also corrupted."
			end
			saxhelper=@saxhelper.new.__init__(listener)
			if source.is_a?(String)
				case @parser
					when :ox           then Ox.sax_parse(saxhelper,StringIO.new(source),:convert_special=>true)
					when :oxhtml       then Ox.sax_parse(saxhelper,StringIO.new(source),:convert_special=>true,:smart=>true)
					when :libxml       then parser=LibXML::XML::SaxParser.string(source);parser.callbacks=saxhelper;parser.parse
					when :nokogiri     then parser=Nokogiri::XML::SAX::Parser.new(saxhelper);parser.parse(source)
					when :nokogirihtml then parser=Nokogiri::HTML::SAX::Parser.new(saxhelper);parser.parse(source)
					when :xmlparser    then saxhelper.parse(source)
					when :oga          then parser=Oga::XML::SaxParser.new(saxhelper,source);parser.parse
					when :ogahtml      then parser=Oga::HTML::SaxParser.new(saxhelper,source);parser.parse
					when :xerces       then parser=XercesR::SAXParser.new;parser.setDocumentHandler(saxhelper);parser.parsebuf(source)
					when :rexmlstream  then REXML::Parsers::StreamParser.new(source,saxhelper).parse
					when :rexmlsax2    then parser=REXML::Parsers::SAX2Parser.new(source);parser.listen(saxhelper);parser.parse
				end
			else
				case @parser
					when :ox           then Ox.sax_parse(saxhelper,source,:convert_special=>true)
					when :oxhtml       then Ox.sax_parse(saxhelper,source,:convert_special=>true,:smart=>true)
					when :libxml       then parser=LibXML::XML::SaxParser.io(source);parser.callbacks=saxhelper;parser.parse
					when :nokogiri     then parser=Nokogiri::XML::SAX::Parser.new(saxhelper);parser.parse(source)
					when :nokogirihtml then parser=Nokogiri::HTML::SAX::Parser.new(saxhelper);parser.parse(source.read) # fixme: nokogirihtml IO doesn't allow errors.
					when :xmlparser    then saxhelper.parse(source)
					when :oga          then parser=Oga::XML::SaxParser.new(saxhelper,source);parser.parse
					when :ogahtml      then parser=Oga::HTML::SaxParser.new(saxhelper,source);parser.parse
					when :xerces       then parser=XercesR::SAXParser.new;parser.setDocumentHandler(saxhelper);parser.parsebuf(source.read)
					when :rexmlstream  then REXML::Parsers::StreamParser.new(source,saxhelper).parse
					when :rexmlsax2    then parser=REXML::Parsers::SAX2Parser.new(source);parser.listen(saxhelper);parser.parse
				end
			end
			listener
		end

		# Parses file as XML. Error handling might be changed in the future.
		def parsefile(filename,listener)
			#begin
				return nil unless FileTest::readable?(filename)
				File.open(filename,'rb'){|f|
					return parse(f,listener)
				}
			#rescue
			#	return nil
			#end
		end
	end

	# This class provides singleton interface to MultiSAX::SAX.
	class Sax
		@@multisax_singleton=SAX.new
		# MultiSAX::SAX#open
		def self.open(*list) @@multisax_singleton.open(*list) end
		# MultiSAX::SAX#reset
		def self.reset() @@multisax_singleton.reset() end
		# MultiSAX::SAX#parser
		def self.parser() @@multisax_singleton.parser() end
		# MultiSAX::SAX#parse
		def self.parse(source,listener) @@multisax_singleton.parse(source,listener) end
		# MultiSAX::SAX#parsefile
		def self.parsefile(filename,listener) @@multisax_singleton.parsefile(filename,listener) end
	end

	# MultiSAX callbacks.
	# MultiSAX::SAX listener should include this module.
	module Callbacks
		# Start of tag
		def sax_tag_start(tag,attrs) end
		# End of tag
		def sax_tag_end(tag) end
		# Comment
		def sax_comment(text) end
		# CDATA
		def sax_cdata(text) end
		# Text
		def sax_text(text) end
		# XML declaration (not available if parser is :libxml)
		def sax_xmldecl(version,encoding,standalone) end
	end
end
